draft-ietf-quic-ack-frequency.md

```diff
@@ -338,25 +351,23 @@ Unreported Missing:
   have not yet been received.
 
 An endpoint that receives an ACK_FREQUENCY frame with a non-zero Reordering
-Threshold value SHOULD send an immediate ACK when the gap
-between the smallest Unreported Missing packet and the Largest Unacked is greater
-than or equal to the Reordering Threshold value. Sending this additional ACK will
-reset the `max_ack_delay` timer and `Ack-Eliciting Threshold` counter as any ACK
-would do.
+Threshold value SHOULD send an immediate ACK in response to an Out-of-Order
+Packet:
 
-In order to ensure timely loss detection, it is optimal to send a Reordering
-Threshold value of 1 less than the packet threshold used by the data sender for
-loss detection. If the threshold is smaller, an ACK_FRAME is sent before the
-packet can be declared lost based on the packet threshold. If the value is
-larger, it causes unnecessary delays. ({{Section 18.2 of QUIC-RECOVERY}})
-recommends a default packet threshold for loss detection of 3, equivalent to
-a Reordering Threshold of 2.
+* when the packet number is larger than the Largest Unacked
+  packet and the difference between the this packet number and the Largest
+  Unacked is larger than the Reordering Threshold.
 
-If the most recent ACK_FREQUENCY frame received from the peer has a `Reordering
-Threshold` value of 0, the endpoint SHOULD NOT send an immediate
-acknowledgement in response to packets received out of order, and instead
-rely on the peer's `Ack-Eliciting Threshold` and `max_ack_delay` thresholds
-for sending acknowledgements.
+* when the packet number is lower than the Largest Acked.
+  (TODO: is that true? Should there be some kind of threshold?)
+
+* when the total number of Unreported Missing packets is larger
+  than the Reordering Threshold.
```

Comment:
I don't think this is needed as it should be covered by case 1

Reply:
The case I am addressing there is:

* threshold is 2.
* the endpoint receives packet 1, no ACK yet because delays. Highest unacked is 1.
* the endpoint receives  packet 3, no ACK yet because the gap to 1 is only one packet.  Highest unacked is 3.
* the endpoint receives packet 5. There is just a one-packet gap to 3, so according to the first rule, no ACK yet. But there are no two holes, 2 and 4, so the endpoint should send an ACK.

Reply:
Come to think of it, we could simplify the specification a lot by just counting the number of unreported missing packets, because rule #3 subsumes rule #1.

Reply:
But then, I don't know whether that's what we actually want to say. Starting from the explanation, we want to send an ACK as soon as we detect a condition that cause the missing packet to be repeated. That condition is "The packet was sent kPacketThreshold packets before an acknowledged packet". That means the endpoint should track the lowest number missing packet number after the highest acked packet, and compare it to the highest unacked. If the difference is larger or equal to the threshold, it should send an acknowledgement. But the text does not say that.

Reply:
Yes, this is actually what the texts tries to say. It's not about the number of missing packets but the only the gap between the last in-order packet and the last to-be-acked packet. The number of missing packet is not important, just the distance.

So in your example above, packet 5 should trigger an ack but also if packet 4 would have arrived in order it should trigger an ack. 

However, assuming your case where packet 4 is missing, you send an ACK when packet 5 arrives because of missing packet 2 (which then is a reported missing packet and does not count for future ack decision anymore). However, as soon as packet 6 arrives, you will send another ack to report of missing packet number 4.